reactor-core/src/main/java21/reactor/core/scheduler/LoomBoundedElasticScheduler.java

```diff
@@ -0,0 +1,1394 @@
+				}
+
+				int a = get();
+				if (!idleQueue.isEmpty()) {
+					//try to find an idle resource
+					BoundedState bs = idleQueue.pollLast();
+					if (bs != null && bs.markPicked()) {
+						boolean accepted = setBusy(bs);
+						if (!accepted) { // shutdown in the meantime
+							bs.shutdown(true);
+							return CREATING;
+						}
+						return bs;
+					}
+					//else optimistically retry (implicit continue here)
+				}
+				else if (a < parent.maxThreads) {
+					//try to build a new resource
+					if (compareAndSet(a, a + 1)) {
+						BoundedState newState = new BoundedState(this);
+						if (newState.markPicked()) {
+							boolean accepted = setBusy(newState);
+							if (!accepted) { // shutdown in the meantime
+								newState.shutdown(true);
+								return CREATING;
+							}
+							return newState;
+						}
+					}
+					//else optimistically retry (implicit continue here)
+				}
+				else {
+					BoundedState s = choseOneBusy();
+					if (s != null && s.markPicked()) {
+						return s;
+					}
+					//else optimistically retry (implicit continue here)
+				}
+			}
+		}
+
+		@Nullable
+		private BoundedState choseOneBusy() {
+			BoundedState[] arr = busyStates.array;
+			int len = arr.length;
+			if (len == 0) {
+				return null; //implicit retry in the pick() loop
+			}
+			if (len == 1) {
+				return arr[0];
+			}
+
+			BoundedState choice = arr[0];
+			int leastBusy = Integer.MAX_VALUE;
+
+			for (int i = 0; i < arr.length; i++) {
+				BoundedState state = arr[i];
+				int busy = state.markCount;
+				if (busy < leastBusy) {
+					leastBusy = busy;
+					choice = state;
+				}
+			}
+			return choice;
+		}
+
+		public BoundedState[] dispose() {
+			BusyStates current;
+			for (;;) {
+				current = busyStates;
+
+				if (current.shutdown) {
+					return current.array;
+				}
+
+				if (BUSY_STATES.compareAndSet(this,
+						current, new BusyStates(current.array, true))) {
+					break;
+				}
+				// the race can happen also with scheduled tasks and eviction
+				// so we need to retry if shutdown transition fails
+			}
+
+			BoundedState[] arr = current.array;
+			// The idleQueue must be drained first as concurrent removals
+			// by evictor or additions by finished tasks can invalidate the size
+			// used if a regular array was created here.
+			// In case of concurrent calls, it is not needed to atomically drain the
+			// queue, nor guarantee same BoundedStates are read, as long as the caller
+			// shuts down the returned BoundedStates. Also, idle ones should easily
+			// shut down, it's not necessary to distinguish graceful from forceful.
+			ArrayList<BoundedState> toAwait = new ArrayList<>(idleQueue.size() + arr.length);
+			BoundedState bs;
+			while ((bs = idleQueue.pollLast()) != null) {
+				toAwait.add(bs);
+			}
+			Collections.addAll(toAwait, arr);
+			return toAwait.toArray(new BoundedState[0]);
+		}
+	}
+
+	static class BoundedState extends CountDownLatch implements Worker, Disposable, Scannable {
```

Comment:
The `BoundedState` probably needs to be decoupled from `Worker` in order to track individual `Worker`'s tasks and their disposal, not to cancel other `Worker`'s tasks if one is disposed.

Reply:
fixed